fix: clarify RetryPollIntervalSeconds/RetryPollCount (#2575)
mendersoftware/mender-docs: 03.Client-installation/07.Configuration/50.Configuration-options/docs.md

#### RetryPollCount

The maximum number of tries that the Mender Client performs when contacting the Mender Server before giving up.

It applies to the following Device APIs:
* [Check Update](https://docs.mender.io/api/#device-api-deployments-v2-check-update)
* [Inventory reporting APIs](https://docs.mender.io/api/#device-api-device-inventory)

> * [Update Deployment Status](https://docs.mender.io/api/#device-api-deployments-update-deployment-status)

Reviewer: Only for two statues, commit and final success/failure

* [Inventory reporting APIs](https://docs.mender.io/api/#device-api-device-inventory)
* [Update Deployment Status](https://docs.mender.io/api/#device-api-deployments-update-deployment-status)
* [Report Deployment Log](https://docs.mender.io/api/#device-api-deployments-report-deployment-log)
* [Download Artifact](https://docs.mender.io/api/#device-api-deployments-download-artifact)

> If the setting is zero (the default), the maximum number of retries is `3 * ceil(log2(RetryPollIntervalSeconds) + 1)`.

Reviewer: This may be still true - I haven't verified it. I don't see a reason to remove it in any case, what the default is otherwise?

#### RetryPollIntervalSeconds

> Interval to retry contacting to the server on failed communications. 
> 
> **Minimum value.** 
> Must be more than 60 seconds. If you specify less than 60, the Mender Client will ignore the setting and use 60 seconds instead.
> 
> **Exponential backoff.**
> The specified interval is used for three consecutive attempts. After three failures, the interval doubles for the next three attempts, and so on, until the total number of attempts (as set by [RetryPollCount](#retrypollcount)) is reached.
> 
> Example for `RetryPollIntervalSeconds=120` and `RetryPollCount=5`:
> 
> ```markdown
> try to connect -> fail
> wait 120sec    -> try to connect -> fail
> wait 120sec    -> try to connect -> fail
> wait 120sec    -> try to connect -> fail
> 
> wait 240sec    -> try to connect -> fail
> wait 240sec    -> try to connect -> fail
> give up
> ```

Reviewer: ```suggestion
Maximum interval to retry contacting to the server on failed communications. 

**Minimum value.** 
Must be more than 60 seconds. If you specify less than 60, the Mender Client will ignore the setting and use 60 seconds instead.

**Exponential backoff.**
The algorithm starts with 60 seconds between retries. After three failures, the interval doubles for the next three attempts, and so on, until reaching  `RetryPollIntervalSeconds`. From there, the interval won't be increased further and it will keep retrying until the total number of attempts (as set by [RetryPollCount](#retrypollcount)) is reached.

Example for `RetryPollIntervalSeconds=300` and `RetryPollCount=15`:

```markdown
try to connect -> fail
wait 60sec    -> try to connect -> fail
wait 60sec    -> try to connect -> fail
wait 60sec    -> try to connect -> fail

wait 120sec    -> try to connect -> fail
wait 120sec    -> try to connect -> fail
wait 120sec    -> try to connect -> fail

wait 240sec    -> try to connect -> fail
wait 240sec    -> try to connect -> fail
wait 240sec    -> try to connect -> fail

wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail

give up
```
```

It applies to the following Device APIs:
* [Check Update](https://docs.mender.io/api/#device-api-deployments-v2-check-update)
* [Inventory reporting APIs](https://docs.mender.io/api/#device-api-device-inventory)

> * [Update Deployment Status](https://docs.mender.io/api/#device-api-deployments-update-deployment-status)

Reviewer: Same as above, only two statues

* [Report Deployment Log](https://docs.mender.io/api/#device-api-deployments-report-deployment-log)

! [Download Artifact](https://docs.mender.io/api/#device-api-deployments-download-artifact) has a fixed retry set to 60 seconds and can't be configured. 

> ! It is **not** affected with the Exponential backoff mechanism, the value remains 60s always.

Reviewer 2: ```suggestion
! It is **not** affected by the Exponential backoff mechanism, the value remains 60s always.
```

> ! It is **not** affected with the Exponential backoff mechanism, the value remains 60s always.

Reviewer: It is yes affected, only that the parameters are harcd-coded to 60 seconds RetryPollIntervalSeconds and 10 times RetryPollCount. But it doubles the same way every three tries.

**Exponential backoff.**
The algorithm starts with 60 seconds between retries. After three failures, the interval doubles for the next three attempts, and so on, until reaching  `RetryPollIntervalSeconds`. From there, the interval won't be increased further and it will keep retrying until the total number of attempts (as set by [RetryPollCount](#retrypollcount)) is reached.

Example for `RetryPollIntervalSeconds=300` and `RetryPollCount=15`:

```markdown
try to connect -> fail
wait 60sec    -> try to connect -> fail
wait 60sec    -> try to connect -> fail
wait 60sec    -> try to connect -> fail

wait 120sec    -> try to connect -> fail
wait 120sec    -> try to connect -> fail
wait 120sec    -> try to connect -> fail

wait 240sec    -> try to connect -> fail
wait 240sec    -> try to connect -> fail
wait 240sec    -> try to connect -> fail

wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail

> wait 300sec (max)    -> try to connect -> fail

Reviewer: Missing \`\`\`

#### RetryPollCount

The maximum number of tries that the Mender Client performs when contacting the Mender Server before giving up.

It applies to the following Device APIs:
* [Check Update](https://docs.mender.io/api/#device-api-deployments-v2-check-update)
* [Inventory reporting APIs](https://docs.mender.io/api/#device-api-device-inventory)
* [Update Deployment Status](https://docs.mender.io/api/#device-api-deployments-update-deployment-status)

>   * Only when reporting status "success" and "failure"

Reviewer: ```suggestion
  * Only when reporting status "commit" and "success" or "failure"
```

Author: Does this behavior make sense to you?

My first opinion is this is a bug and not something to document.

Reviewer: In my opinion it is a limitation on the API (because there is no "committing" state) that the client is doing a workaround for. It is a very important step for the Client to try contacting the server just before committing to detect for example aborted deployments that must not be committed. That is why we retry here and treat the failure as a fatal one (note that the status updates that are not retried are not treated as failures, the client goes on).

We shall document that during commit we use this algorithm to submit the status, we don't need to document which exact string are we sending to do so.

Example for `RetryPollIntervalSeconds=300` and `RetryPollCount=15`:

```markdown
try to connect -> fail
wait 60sec    -> try to connect -> fail
wait 60sec    -> try to connect -> fail
wait 60sec    -> try to connect -> fail

wait 120sec    -> try to connect -> fail
wait 120sec    -> try to connect -> fail
wait 120sec    -> try to connect -> fail

wait 240sec    -> try to connect -> fail
wait 240sec    -> try to connect -> fail
wait 240sec    -> try to connect -> fail

wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail
wait 300sec (max)    -> try to connect -> fail

give up
```

It applies to the following Device APIs:
* [Check Update](https://docs.mender.io/api/#device-api-deployments-v2-check-update)
* [Inventory reporting APIs](https://docs.mender.io/api/#device-api-device-inventory)
* [Update Deployment Status](https://docs.mender.io/api/#device-api-deployments-update-deployment-status)

>   * Only when reporting status "success" and "failure"

Reviewer: ```suggestion
  * Only when reporting status "commit" and "success" or "failure"
```

> ! [Download Artifact](https://docs.mender.io/api/#device-api-deployments-download-artifact) has a fixed retry duration of 60 seconds and can't be configured. 

Reviewer: This is dangling here with no context. From the list above there is no mention of Download. Also, this has nothing to do with the Artifact download **API**, that is never retried. And to be clear, this is not configurable so not really related to this docs page at all.

Based on the previous reasoning **I suggest to remove it**. If anything, I would write: "The same algorithm is also applied when resuming an Artifact download: when the connection with the server is lost in the middle of a Download, the client will resume from the last received byte using the same algorithm explained above, but with a fixed retry duration of 60 seconds that cannot be configured"

Also, style:

```suggestion
!!! [Download Artifact](https://docs.mender.io/api/#device-api-deployments-download-artifact) has a fixed retry duration of 60 seconds and can't be configured. 
```

:point_up: It does not need to be a warning, IMO an info note (blue) is enough.

Author: > Also, this has nothing to do with the Artifact download API, that is never retried. 

Let me see if I got this.

Calling the "API do get the download link from s3" ([GET /artifacts/{id}/download](https://docs.mender.io/api/#device-api-deployments-download-artifact)) has the fixed 60sec duration.

Downloading the artifact from the artifact storage (i.e https://s3.amazonaws.com/hosted-mender-artifacts), which isn't an API call,  has no retries. 

Is this correct?

Reviewer: > Is this correct?

No. Calling the API has no retries. When downloading, we have a separate feature that we call "download resume" so that I you already downloaded a part of the Artifact and you loose the connection, you resume a bit later (in 60 seconds) using HTTP Range requests to skip the already downloaded data.

So, in other words, it has nothing to do in my opinion with the configuration settings that we are documenting in this page.
